hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowRollbacksProcedure.scala

```diff
@@ -0,0 +1,147 @@
+  ))
+
+  def parameters: Array[ProcedureParameter] = if (showRollback) ROLLBACK_PARAMETERS else ROLLBACKS_PARAMETERS
+
+  def outputType: StructType = if (showRollback) ROLLBACK_OUTPUT_TYPE else ROLLBACKS_OUTPUT_TYPE
+
+  override def call(args: ProcedureArgs): Seq[Row] = {
+    super.checkArgs(parameters, args)
+
+    val tableName = getArgValueOrDefault(args, parameters(0))
+    val limit = getArgValueOrDefault(args, parameters(1)).get.asInstanceOf[Int]
+
+    val basePath = getBasePath(tableName)
+    val metaClient = HoodieTableMetaClient.builder.setConf(jsc.hadoopConfiguration()).setBasePath(basePath).build
+    val activeTimeline = new RollbackTimeline(metaClient)
+    if (showRollback) {
+      val instantTime = getArgValueOrDefault(args, parameters(2)).get.asInstanceOf[String]
+      getRollback(activeTimeline, instantTime, limit)
+    } else {
+      getRollbacks(activeTimeline, limit)
+    }
+  }
+
+  override def build: Procedure = new ShowRollbacksProcedure(showRollback)
+
+  class RollbackTimeline(metaClient: HoodieTableMetaClient) extends HoodieActiveTimeline(metaClient,
+    CollectionUtils.createImmutableSet(HoodieTimeline.ROLLBACK_EXTENSION)) {
+  }
+
+  def getRollback(activeTimeline: RollbackTimeline,
+                  instantTime: String,
+                  limit: Int): Seq[Row] = {
+    val rows = new util.ArrayList[Row]
+    val metadata = TimelineMetadataUtils.deserializeAvroMetadata(activeTimeline.getInstantDetails(
+      new HoodieInstant(State.COMPLETED, ROLLBACK_ACTION, instantTime)).get, classOf[HoodieRollbackMetadata])
+
+    metadata.getPartitionMetadata.asScala.toMap.iterator.foreach(entry => Stream
+      .concat(entry._2.getSuccessDeleteFiles.map(f => Pair.of(f, true)),
+        entry._2.getFailedDeleteFiles.map(f => Pair.of(f, false)))
+      .foreach((fileWithDeleteStatus: Pair[String, Boolean]) => {
+        rows.add(Row(metadata.getStartRollbackTime, metadata.getCommitsRollback.toString,
+          entry._1, fileWithDeleteStatus.getLeft, fileWithDeleteStatus.getRight))
+      }))
+    rows.stream().limit(limit).toArray().map(r => r.asInstanceOf[Row]).toList
+  }
+
+  def getRollbacks(activeTimeline: RollbackTimeline,
+                   limit: Int): Seq[Row] = {
+    val rows = new util.ArrayList[Row]
+    val rollback = activeTimeline.getRollbackTimeline.filterCompletedInstants
+
+    rollback.getInstants.iterator().asScala.foreach(instant => {
+      try {
+        val metadata = TimelineMetadataUtils.deserializeAvroMetadata(activeTimeline.getInstantDetails(instant).get,
+          classOf[HoodieRollbackMetadata])
+
+        metadata.getCommitsRollback.iterator().asScala.foreach(c => {
+          rows.add(Row(metadata.getStartRollbackTime, c,
+            metadata.getTotalFilesDeleted, metadata.getTimeTakenInMillis,
+            if (metadata.getPartitionMetadata != null) metadata.getPartitionMetadata.size else 0))
+        })
+      } catch {
+        case e: IOException =>
+          e.printStackTrace()
```

Comment:
why eat exception?

Reply:
This is implemented in `RollbacksCommand` class in hudi-cli, I think it is to prevent the IOException of an instant from causing other instants to get.